RabbitMQ Basic Consumer support for sequential processing with ConcurrentMessageLimit = 1

sam9291 opened the pull request:
This is an implementation proposal to support for RabbitMQ consumer to process messages sequentially when PrefetchCount is set to 0 or 2+ and `RabbitMqReceiveEndpointContext.ConcurrentMessageLimit` is set to 1 to avoid concurrent message processing.

In some scenario, such capability can be useful. 

See the integration test `Using_a_consumer_concurrency_limit_set_to_1.Should_limit_the_consumer_and_consume_messages_sequentially()` method for a concrete example.

Here is a visual summary of the change:

### Before

![image](https://github.com/MassTransit/MassTransit/assets/5232883/b88fc21b-d2f1-4b12-96fa-733eb4e805af)

### After

![image](https://github.com/MassTransit/MassTransit/assets/5232883/a18e1026-d496-45db-b7b1-70db2b607035)

phatboyg commented:
So, this likely won't be accepted as there are faster primitives already built into MassTransit to support this.

@NooNameR can offer some guidance.

phatboyg commented:
I think SQS and Kafka both use it, to ensure in-order delivery within the same partition key/group id.

sam9291 commented:
Nice, I see the `FifoChannelExecutorPool` in the SQA receiver, I'll rework this PR to align with this design, thanks for the heads up!

NooNameR commented:
for the thing you are doing, you need only `new ChannelExecutor(prefetchCount, 1)`:
https://github.com/MassTransit/MassTransit/blob/1f13a70aa02effb17b7784c2d4d9f7350623b993/src/MassTransit/Util/ChannelExecutor.cs#L96

I do not think you want distribute it across partitions as in case of SQS or Kafka

sam9291 commented:
Correct, I stumbled upon it with the my team, we started the implementation with `ChannelExecutor`, makes much more sense with that :D 

However I don't see a way yet to not set the ConcurrentMessageLimit in case it was not configure for the endpoint (`!context.ConcurrentMessageLimit.HasValue`) currently looking into it

NooNameR commented:
> Correct, I stumbled upon it with the my team, we started the implementation with `ChannelExecutor`, makes much more sense with that :D
> 
> However I don't see a way yet to not set the ConcurrentMessageLimit in case it was not configure for the endpoint (`!context.ConcurrentMessageLimit.HasValue`) currently looking into it

well, programming is tough 😄 
you can modify channel executor to not create semaphore, maybe? or just use separate approaches.

Basically to understand the end goal, you want to fix ordering?

sam9291 commented:
Exactly, my main goal is that when we set the concurrent message limit to 1, I would expect it to process messages in the order RabbitMq delivers them. I need this for a multiple use cases on a project, but I can imagine it being useful to others as well.

To minimize the channel executor changes, this approach seems to be working great:

![image](https://github.com/MassTransit/MassTransit/assets/5232883/06f03385-7314-4551-9a3a-2fb3a515813c)

NooNameR commented:
we just chat with @phatboyg and it just requires to move `_limit.Wait` before `Task.Run`:

https://github.com/MassTransit/MassTransit/blob/a04955c2712c565612c8da3c19e8fa38d2bcd52d/src/Transports/MassTransit.RabbitMqTransport/RabbitMqTransport/RabbitMqBasicConsumer.cs#L164-L165

sam9291 commented:
We can now use my PR as over-engineering example, how did I miss that 😂. Updating and testing this rightaway!

phatboyg commented:
IF, and ONLY IF, ConcurrentMessageLimit = 1

sam9291 commented on tests/MassTransit.RabbitMqTransport.Tests/ConcurrencyFilter_Specs.cs:
```diff
@@ -91,5 +91,111 @@ class A
+
+                Interlocked.Decrement(ref _currentPendingDeliveryCount);
+
+                _completedConsumingSequentially = _previousSequenceIndex == sequenceIndex - 1;
```
Problem here I need to fix, this only check if the last two messages were processed in sequence

sam9291 reviewed (commented):
Fixed

sam9291 commented:
> would be nice if you can run a benchmark for that. we have an internal tool for that.
> 
> https://github.com/MassTransit/MassTransit/blob/develop/tests/MassTransit.Benchmark/Program.cs
> 
> but not only for concurrent limit 1, we need to check how this affects > 1

Here is the benchmark results I got with the latest changes of this PR:

<table>
<tr>
<td> Before </td> <td> After </td>
</tr>
<tr>
<td> 


```
MassTransit Benchmark

Transport: RabbitMq
Operating System: Microsoft Windows NT 10.0.22621.0
Processor Count: 20
MassTransit Version: 1.0.0.0
Host: localhost
Virtual Host: /
Username: guest
Password: *****
Heartbeat: 00:00:00
Publisher Confirmation: False
Split: False
Batch: enabled=True, limit=100, timeout=1ms
Running Message Latency Benchmark
Message Count: 10000
Clients: 10
Durable: False
Payload Length: 0
Prefetch Count: 100
Concurrency Limit: 0
Total send duration: 0:00:15.3461925
Send message rate: 651.63 (msg/s)
Total consume duration: 0:00:15.3648832
Consume message rate: 650.83 (msg/s)
Concurrent Consumer Count: 20
Avg Ack Time: 15ms
Min Ack Time: 0ms
Max Ack Time: 79ms
Med Ack Time: 15ms
95t Ack Time: 16ms
Avg Consume Time: 39ms
Min Consume Time: 3ms
Max Consume Time: 110ms
Med Consume Time: 39ms
95t Consume Time: 63ms

    3ms ****                                                         (    158)
   14ms ************************************************************ (   2263)
   25ms ****************************************************         (   1970)
   35ms *****************************************************        (   2028)
   46ms **********************************************************   (   2188)
   57ms **********************************                           (   1313)
   67ms *                                                            (     70)
   99ms                                                              (      9)
Host: localhost
Virtual Host: /
Username: guest
Password: *****
Heartbeat: 00:00:00
Publisher Confirmation: False
Split: False
Batch: enabled=True, limit=100, timeout=1ms
Running Request Response Benchmark
Message Count: 10000
Clients: 10
Durable: False
Prefetch Count: 100
Concurrency Limit: 0
Total consume duration: 0:00:46.8229281
Consume message rate: 213.57 (msg/s)
Total request duration: 0:00:46.8334682
Request rate: 213.52 (msg/s)
Concurrent Consumer Count: 10
Avg Request Time: 46ms
Min Request Time: 3ms
Max Request Time: 114ms
Med Request Time: 48ms
95t Request Time: 65ms
Avg Consume Time: 31ms
Min Consume Time: 1ms
Max Consume Time: 81ms
Med Consume Time: 42ms
95t Consume Time: 52ms

Request duration distribution
    3ms **                                                           (    191)
   14ms *********                                                    (    668)
   25ms ********************************************************     (   3816)
   36ms ****                                                         (    295)
   47ms *********                                                    (    649)
   58ms ************************************************************ (   4083)
   69ms **                                                           (    162)
   92ms *                                                            (    115)
  103ms                                                              (     20)
```

</td>

<td> 


```
MassTransit Benchmark

Transport: RabbitMq
Operating System: Microsoft Windows NT 10.0.22621.0
Processor Count: 20
MassTransit Version: 1.0.0.0
Host: localhost
Virtual Host: /
Username: guest
Password: *****
Heartbeat: 00:00:00
Publisher Confirmation: False
Split: False
Batch: enabled=True, limit=100, timeout=1ms
Running Message Latency Benchmark
Message Count: 10000
Clients: 10
Durable: False
Payload Length: 0
Prefetch Count: 100
Concurrency Limit: 0
Total send duration: 0:00:15.9507997
Send message rate: 626.93 (msg/s)
Total consume duration: 0:00:16.0001556
Consume message rate: 624.99 (msg/s)
Concurrent Consumer Count: 20
Avg Ack Time: 15ms
Min Ack Time: 0ms
Max Ack Time: 164ms
Med Ack Time: 15ms
95t Ack Time: 16ms
Avg Consume Time: 41ms
Min Consume Time: 3ms
Max Consume Time: 196ms
Med Consume Time: 40ms
95t Consume Time: 63ms

    3ms ************************                                     (   1589)
   23ms ************************************************************ (   3919)
   42ms *********************************************************    (   3773)
   61ms **********                                                   (    679)
  100ms                                                              (     10)
  157ms                                                              (     10)
  177ms                                                              (     19)
Host: localhost
Virtual Host: /
Username: guest
Password: *****
Heartbeat: 00:00:00
Publisher Confirmation: False
Split: False
Batch: enabled=True, limit=100, timeout=1ms
Running Request Response Benchmark
Message Count: 10000
Clients: 10
Durable: False
Prefetch Count: 100
Concurrency Limit: 0
Total consume duration: 0:00:50.0258178
Consume message rate: 199.90 (msg/s)
Total request duration: 0:00:50.0444945
Request rate: 199.82 (msg/s)
Concurrent Consumer Count: 10
Avg Request Time: 49ms
Min Request Time: 2ms
Max Request Time: 127ms
Med Request Time: 60ms
95t Request Time: 65ms
Avg Consume Time: 33ms
Min Consume Time: 1ms
Max Consume Time: 108ms
Med Consume Time: 43ms
95t Consume Time: 52ms

Request duration distribution
    2ms **                                                           (    243)
   15ms ******                                                       (    586)
   27ms *********************************                            (   2811)
   40ms ********                                                     (    687)
   52ms ************************************************************ (   5056)
   64ms *****                                                        (    450)
   77ms                                                              (     28)
   89ms *                                                            (    126)
  102ms                                                              (      7)
  114ms                                                              (      5)
```

</td>

</tr>

<tr>

<td>

```
MassTransit Benchmark

Transport: RabbitMq
Operating System: Microsoft Windows NT 10.0.22621.0
Processor Count: 20
MassTransit Version: 1.0.0.0
Host: localhost
Virtual Host: /
Username: guest
Password: *****
Heartbeat: 00:00:00
Publisher Confirmation: False
Split: False
Batch: enabled=True, limit=100, timeout=1ms
Running Message Latency Benchmark
Message Count: 10000
Clients: 10
Durable: False
Payload Length: 0
Prefetch Count: 100
Concurrency Limit: 1
Total send duration: 0:00:15.6461456
Send message rate: 639.14 (msg/s)
Total consume duration: 0:00:15.6494052
Consume message rate: 639.00 (msg/s)
Concurrent Consumer Count: 1
Avg Ack Time: 15ms
Min Ack Time: 0ms
Max Ack Time: 66ms
Med Ack Time: 15ms
95t Ack Time: 16ms
Avg Consume Time: 35ms
Min Consume Time: 2ms
Max Consume Time: 97ms
Med Consume Time: 36ms
95t Consume Time: 53ms

    2ms **                                                           (    115)
   12ms *******************************************                  (   2306)
   21ms *********************                                        (   1172)
   31ms ************************************************************ (   3216)
   40ms ******************                                           (    997)
   50ms ****************************************                     (   2175)
   59ms                                                              (      9)
   88ms                                                              (      9)
Host: localhost
Virtual Host: /
Username: guest
Password: *****
Heartbeat: 00:00:00
Publisher Confirmation: False
Split: False
Batch: enabled=True, limit=100, timeout=1ms
Running Request Response Benchmark
Message Count: 10000
Clients: 10
Durable: False
Prefetch Count: 100
Concurrency Limit: 1
Total consume duration: 0:00:48.3952889
Consume message rate: 206.63 (msg/s)
Total request duration: 0:00:48.41218
Request rate: 206.56 (msg/s)
Concurrent Consumer Count: 1
Avg Request Time: 48ms
Min Request Time: 3ms
Max Request Time: 133ms
Med Request Time: 59ms
95t Request Time: 66ms
Avg Consume Time: 31ms
Min Consume Time: 1ms
Max Consume Time: 89ms
Med Consume Time: 41ms
95t Consume Time: 53ms

Request duration distribution
    3ms ******                                                       (    502)
   16ms **********                                                   (    875)
   29ms ************************************                         (   2947)
   42ms *****                                                        (    415)
   55ms ************************************************************ (   4855)
   68ms *                                                            (    161)
   81ms                                                              (     38)
   94ms **                                                           (    191)
  107ms                                                              (      7)
  120ms                                                              (      8)
```

</td>

<td>

```
MassTransit Benchmark

Transport: RabbitMq
Operating System: Microsoft Windows NT 10.0.22621.0
Processor Count: 20
MassTransit Version: 1.0.0.0
Host: localhost
Virtual Host: /
Username: guest
Password: *****
Heartbeat: 00:00:00
Publisher Confirmation: False
Split: False
Batch: enabled=True, limit=100, timeout=1ms
Running Message Latency Benchmark
Message Count: 10000
Clients: 10
Durable: False
Payload Length: 0
Prefetch Count: 100
Concurrency Limit: 1
Total send duration: 0:00:15.7906008
Send message rate: 633.29 (msg/s)
Total consume duration: 0:00:15.8347839
Consume message rate: 631.52 (msg/s)
Concurrent Consumer Count: 1
Avg Ack Time: 15ms
Min Ack Time: 0ms
Max Ack Time: 86ms
Med Ack Time: 16ms
95t Ack Time: 16ms
Avg Consume Time: 36ms
Min Consume Time: 5ms
Max Consume Time: 120ms
Med Consume Time: 36ms
95t Consume Time: 53ms

    5ms ****                                                         (    217)
   17ms ************************************************************ (   3235)
   28ms **********************************************************   (   3166)
   40ms *****************************************                    (   2254)
   51ms ********************                                         (   1118)
  109ms                                                              (      9)
Host: localhost
Virtual Host: /
Username: guest
Password: *****
Heartbeat: 00:00:00
Publisher Confirmation: False
Split: False
Batch: enabled=True, limit=100, timeout=1ms
Running Request Response Benchmark
Message Count: 10000
Clients: 10
Durable: False
Prefetch Count: 100
Concurrency Limit: 1
Total consume duration: 0:00:50.1942994
Consume message rate: 199.23 (msg/s)
Total request duration: 0:00:50.2046088
Request rate: 199.18 (msg/s)
Concurrent Consumer Count: 1
Avg Request Time: 50ms
Min Request Time: 4ms
Max Request Time: 114ms
Med Request Time: 60ms
95t Request Time: 68ms
Avg Consume Time: 33ms
Min Consume Time: 1ms
Max Consume Time: 78ms
Med Consume Time: 44ms
95t Consume Time: 53ms

Request duration distribution
    4ms ***                                                          (    281)
   15ms *******                                                      (    647)
   26ms ************************************                         (   2997)
   37ms *                                                            (    143)
   48ms ******                                                       (    535)
   59ms ************************************************************ (   4985)
   70ms *                                                            (     99)
   92ms ***                                                          (    271)
  103ms                                                              (     41)
```

</td>

</tr>

<tr>

<td>

```
MassTransit Benchmark

Transport: RabbitMq
Operating System: Microsoft Windows NT 10.0.22621.0
Processor Count: 20
MassTransit Version: 1.0.0.0
Host: localhost
Virtual Host: /
Username: guest
Password: *****
Heartbeat: 00:00:00
Publisher Confirmation: False
Split: False
Batch: enabled=True, limit=100, timeout=1ms
Running Message Latency Benchmark
Message Count: 10000
Clients: 10
Durable: False
Payload Length: 0
Prefetch Count: 100
Concurrency Limit: 2
Total send duration: 0:00:15.6777259
Send message rate: 637.85 (msg/s)
Total consume duration: 0:00:15.6928454
Consume message rate: 637.23 (msg/s)
Concurrent Consumer Count: 2
Avg Ack Time: 15ms
Min Ack Time: 0ms
Max Ack Time: 133ms
Med Ack Time: 15ms
95t Ack Time: 16ms
Avg Consume Time: 35ms
Min Consume Time: 3ms
Max Consume Time: 168ms
Med Consume Time: 36ms
95t Consume Time: 53ms

    3ms ************************                                     (   1688)
   20ms ************************************************************ (   4207)
   36ms ******************************************************       (   3855)
   53ms ***                                                          (    240)
  151ms                                                              (      9)
Host: localhost
Virtual Host: /
Username: guest
Password: *****
Heartbeat: 00:00:00
Publisher Confirmation: False
Split: False
Batch: enabled=True, limit=100, timeout=1ms
Running Request Response Benchmark
Message Count: 10000
Clients: 10
Durable: False
Prefetch Count: 100
Concurrency Limit: 2
Total consume duration: 0:00:48.182545
Consume message rate: 207.54 (msg/s)
Total request duration: 0:00:48.1861075
Request rate: 207.53 (msg/s)
Concurrent Consumer Count: 2
Avg Request Time: 48ms
Min Request Time: 4ms
Max Request Time: 107ms
Med Request Time: 57ms
95t Request Time: 68ms
Avg Consume Time: 31ms
Min Consume Time: 1ms
Max Consume Time: 82ms
Med Consume Time: 42ms
95t Consume Time: 52ms

Request duration distribution
    4ms **                                                           (    197)
   14ms *********                                                    (    699)
   25ms **********************************************               (   3478)
   35ms                                                              (     71)
   45ms ******                                                       (    492)
   56ms ************************************************************ (   4490)
   66ms ***                                                          (    274)
   76ms                                                              (     64)
   87ms *                                                            (    100)
   97ms *                                                            (    134)
```

</td>

<td>

```
MassTransit Benchmark

Transport: RabbitMq
Operating System: Microsoft Windows NT 10.0.22621.0
Processor Count: 20
MassTransit Version: 1.0.0.0
Host: localhost
Virtual Host: /
Username: guest
Password: *****
Heartbeat: 00:00:00
Publisher Confirmation: False
Split: False
Batch: enabled=True, limit=100, timeout=1ms
Running Message Latency Benchmark
Message Count: 10000
Clients: 10
Durable: False
Payload Length: 0
Prefetch Count: 100
Concurrency Limit: 2
Total send duration: 0:00:15.116871
Send message rate: 661.51 (msg/s)
Total consume duration: 0:00:15.1490963
Consume message rate: 660.11 (msg/s)
Concurrent Consumer Count: 2
Avg Ack Time: 14ms
Min Ack Time: 0ms
Max Ack Time: 133ms
Med Ack Time: 15ms
95t Ack Time: 16ms
Avg Consume Time: 35ms
Min Consume Time: 3ms
Max Consume Time: 165ms
Med Consume Time: 37ms
95t Consume Time: 53ms

    3ms **********                                                   (    730)
   19ms *******************************************************      (   3767)
   35ms ************************************************************ (   4041)
   51ms *********************                                        (   1452)
  149ms                                                              (      9)
Host: localhost
Virtual Host: /
Username: guest
Password: *****
Heartbeat: 00:00:00
Publisher Confirmation: False
Split: False
Batch: enabled=True, limit=100, timeout=1ms
Running Request Response Benchmark
Message Count: 10000
Clients: 10
Durable: False
Prefetch Count: 100
Concurrency Limit: 2
Total consume duration: 0:00:50.5640497
Consume message rate: 197.77 (msg/s)
Total request duration: 0:00:50.5812163
Request rate: 197.70 (msg/s)
Concurrent Consumer Count: 2
Avg Request Time: 50ms
Min Request Time: 6ms
Max Request Time: 130ms
Med Request Time: 60ms
95t Request Time: 69ms
Avg Consume Time: 33ms
Min Consume Time: 1ms
Max Consume Time: 75ms
Med Consume Time: 44ms
95t Consume Time: 52ms

Request duration distribution
    6ms ******                                                       (    540)
   18ms ********************                                         (   1814)
   31ms **************                                               (   1303)
   43ms ******                                                       (    554)
   56ms ************************************************************ (   5284)
   68ms **                                                           (    216)
   81ms                                                              (     30)
   93ms **                                                           (    258)
```

</td>

</tr>
</table>
